File: 02. Algorithms/01. Arrays/08. Kadane's Algorithm/918. Maximum Sum Circular Subarray/Solution.java
// https://leetcode.com/problems/maximum-sum-circular-subarray/
class Solution {
    public int maxSubarraySumCircular(int[] arr) {
        int n = arr.length;
        int allsum = 0;
        for(int ele: arr)
            allsum += ele;
        
        int sumA = maximumSumSA(arr, n);
        int sumB = allsum - minimumSumSA(arr, n);
        if(sumA > 0)
            return Math.max(sumA, sumB);
        else
            return sumA;
    }
    public int maximumSumSA(int[] arr, int n) {
        int csum = arr[0]; //current sum
        int osum = arr[0]; //overall sum
        for(int i = 1; i < n; i++) {
            if(csum > 0)
                csum += arr[i];
            else
                csum = arr[i];
            
            if(csum > osum)
                osum = csum;
        }
        return osum;
    }
    public int minimumSumSA(int[] arr, int n) {
        int csum = arr[0]; //current sum
        int osum = arr[0]; //overall sum
        for(int i = 1; i < arr.length; i++) {
            if(csum < 0)
                csum += arr[i];
            else
                csum = arr[i];
            
            if(csum < osum)
                osum = csum;
        }
        return osum;
    }
}

File: 02. Algorithms/01. Arrays/08. Kadane's Algorithm/918. Maximum Sum Circular Subarray/Solution.js
/**
 * https://leetcode.com/problems/maximum-sum-circular-subarray/
 * Time O(n) | Space O(1)
 * @param {number[]} nums
 * @return {number}
 */
var maxSubarraySumCircular = function(nums) {
    
    let currMin  = 0;
    let globleMin = nums[0];
    let currMax   = 0;
    let globleMax = nums[0];

    let total = 0;
    for(let i = 0; i < nums.length; i++) {
        currMin = Math.min(currMin + nums[i], nums[i]);
        globleMin = Math.min(currMin, globleMin);
        currMax = Math.max(currMax + nums[i], nums[i]);
        globleMax = Math.max(currMax, globleMax);
        total += nums[i]; 
    }

    if(globleMax < 0) {
        return globleMax;
    }

    return Math.max(globleMax, total - globleMin);
};
